compiler/damlc/daml-stdlib-src/DA/Internal/Desugar.daml

```diff
@@ -35,4 +35,19 @@ data NonConsuming t = NonConsuming {}
 data PreConsuming t = PreConsuming {}
 data Consuming t = Consuming {}
 data PostConsuming t = PostConsuming {}
-data Implements t i = Implements
+
+class Implements t i where
+  toInterface : t -> i
+  toInterface = GHC.Types.primitive @"EToInterface"
+
+  fromInterface : i -> Optional t
+  fromInterface = GHC.Types.primitive @"EFromInterface"
+
+  toInterfaceContractId : ContractId t -> ContractId i
+  toInterfaceContractId = GHC.Types.primitive @"EToInterfaceContractId"
+
+  fromInterfaceContractId : ContractId i -> Update (Optional (ContractId t))
+  fromInterfaceContractId = GHC.Types.primitive @"UFromInterfaceContractId"
+
+  interfaceTypeRep : forall p. p i -> t -> TypeRep
```

Comment:
I added the `p i` argument since otherwise the type of the method would be ambiguous. I first thought of using `Proxy` but it seems that's not a thing in Daml. Alternatively, we could just enable `AllowAmbiguousTypes` here, but I thought I'd rather avoid the extension here.

Reply:
We use AllowAmbiguousTypes in a bunch of places so I wouldn’t really worry about that. As long as users don’t have to enable it, you should be good.

Reply:
Fair! I just put the old type back now